src/rgw/rgw_rest.cc

```diff
@@ -922,15 +930,92 @@ int RGWPutObj_ObjStore::get_data(bufferlist& bl)
   }
 
   int len = 0;
-  if (cl) {
-    bufferptr bp(cl);
+  if (cl) 
+  {
+    if (kmsdata)
+    {
+      uint64_t buffer_length = cl;
+      if (cl == 16)
+        buffer_length = 32;
+      else if (cl< 16)
+        buffer_length = 16;
+      bufferptr bp(buffer_length);
+
+      int read_len; /* cio->read() expects int * */
+      int r = s->cio->read(bp.c_str(), cl, &read_len);
+      if (r < 0)
+        return r;
+      unsigned char* read_data = reinterpret_cast<unsigned char *>(bp.c_str());
+      if (hash && read_len)
+        hash->Update((const byte *)bp.c_str(),read_len);
+
+      dout(0) << "gbdebug Earlier Read length " << read_len << " " << kmsdata->key_dec << " is" << dendl;
+      if (read_len < 16 && read_len > 0)
+      {
+        unsigned deficit = 16 - read_len;
+        s->content_length += deficit;
+        dout(0) << "gbdebug deficit " << deficit << " " << dendl;
+        for (int i = read_len ; i < 16 ; i++)
+          read_data[i] = deficit + '0'; //for now
+        read_len = 16; 
+      }
+      else if (read_len == 16 && read_len > 0)
+      {
+        s->content_length += 16;
+        for (int i = read_len; i < 32; i++)
+          read_data[i] = '0'; //for now
+        read_len = 32; 
+      }
+      len = read_len;
+
+      /* Initialise the library */
+      ERR_load_crypto_strings();
+      OpenSSL_add_all_algorithms();
+      OPENSSL_config(NULL);
+
+      unsigned char ciphertext[2*read_len];
+      //unsigned char decryptedtext[2*read_len];
```

Comment:
why 2*read_len ? Also, dont allocate on the stack. too many put requests can lead to stack overflow.